src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,291 @@
+                _sortOption = value;
+            }
+        }
+
+        private SortOption _sortOption;
+
+        /// <summary>
+        /// Queue to batch up work to do to compute the data model. Used so we can batch up a lot of events 
+        /// and only fetch the model once for every batch. The bool parameter is unused.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?> _computeDataModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to filter and sort the data model. The bool parameter is unused.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?> _filterAndSortDataModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to highlight the currently selected symbol node, expand/collapse nodes,
+        /// then update the UI.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<ExpansionOption> _highlightExpandAndPresentItemsQueue;
+
+        /// <summary>
+        /// Keeps track of the current primary and secondary text views. Should only be accessed by the UI thread.
+        /// </summary>
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
+
+        public DocumentOutlineControl(
+            ILanguageServiceBrokerShim languageServiceBroker,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            InitializeComponent();
+
+            _languageServiceBroker = languageServiceBroker;
+            _threadingContext = threadingContext;
+            _asyncListener = asyncListener;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _codeWindow = codeWindow;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _cancellationToken = _cancellationTokenSource.Token;
+            SortOption = SortOption.Location;
+
+            _computeDataModelQueue = new AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?>(
+                DelayTimeSpan.Short,
+                ComputeDataModelAsync,
+                EqualityComparer<bool>.Default,
+                asyncListener,
+                _cancellationToken);
+
+            _filterAndSortDataModelQueue = new AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?>(
+                DelayTimeSpan.NearImmediate,
+                FilterAndSortDataModelAsync,
+                EqualityComparer<bool>.Default,
+                asyncListener,
+                _cancellationToken);
+
+            _highlightExpandAndPresentItemsQueue = new AsyncBatchingWorkQueue<ExpansionOption>(
+                DelayTimeSpan.NearImmediate,
+                HighlightExpandAndPresentItemsAsync,
+                asyncListener,
+                _cancellationToken);
```

Comment:
note to self: this should only look at the last ExpansionOption in teh array when it actually processes its work.